torch/nn/modules/module.py

```diff
@@ -189,17 +190,35 @@ def add_module(self, name, module):
             raise KeyError("module name can't be empty string \"\"")
         self._modules[name] = module
 
-    def _apply(self, fn):
+    def _apply(self, fn, update_params_inplace=True):
         for module in self.children():
-            module._apply(fn)
+            module._apply(fn, update_params_inplace)
 
-        for param in self._parameters.values():
+        for key, param in self._parameters.items():
             if param is not None:
-                # Tensors stored in modules are graph leaves, and we don't
-                # want to create copy nodes, so we have to unpack the data.
-                param.data = fn(param.data)
-                if param._grad is not None:
-                    param._grad.data = fn(param._grad.data)
+                param_applied = fn(param.data)
+                if not update_params_inplace:
+                    # Tensors stored in modules are graph leaves, and we don't want to
+                    # create copy nodes, so we have to use `with torch.no_grad():`
+                    with torch.no_grad():
+                        self.register_parameter(key, Parameter(param_applied, param.requires_grad))
+                        # Bump up the version counter of the original parameter, to invalidate
+                        # any previous references of it in the autograd graph.
+                        param.add_(0.0)
```

Comment:
can we just call an internal method (could be manually implemented on Variable) that does this?  The issue here is that you'll trace `add_`, which is bogus.